### PR TITLE
Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by MarcoDiaz2000**
## Our team is formed by:
Team member # 1: @ab-noori 
Team member # 2: @MarcoDiaz2000
Team member # 3: @tamoorsaeed22 

## Description

In this project, we create a console application that will help you keep track of the different types of things you own: books, music albums and games.
Everything will be based on the UML class diagram below. The data will be stored in JSON files, but you will also prepare a database with a table structure analogous to the class structure of your program.

catalog_of_my_things1

#### Logic

- Each class from the [UML class diagram](./images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. _Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created._
- All methods visible in the diagram should be implemented:
    - add_item method (in all classes that have that method)
        - should take an instance of the Item class as an input.
        - should add the input item to the collection of items.
        - should add self as a property of the item object (by using the correct setter from the item object).
    - can_be_archived?() in the Item class
        - should return true if published_date is older than 10 years.
        - otherwise, it should return false.
    - move_to_archive() in the Item class
        - should reuse can_be_archived?() method.
        - should change the archived property to true if the result of the can_be_archived?() method is true.
        - should do nothing if the result of the can_be_archived?() method is false.
    - can_be_archived?() in the Book class
        - should override the method from the parent class.
        - should return true if parent's method returns true OR if cover_state equals to "bad".
        - otherwise, it should return false.
    - can_be_archived?() in the MusicAlbum class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if on_spotify equals true.
        - otherwise, it should return false.
    - can_be_archived?() in the Game class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if last_played_at is older than 2 years.
        - otherwise, it should return false.
- Add unit tests for all implemented methods.

#### User interface

- Create a main.rb file that will serve as your console app entry-point.
- Your console app, at the start, should:
    - Present the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
    - Have a way to quit the app.
- The following options should be available:
    - List all books
    - List all music albums
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - Add a book
    - Add a music album
    - Add a game
 - All data should be preserved by saving collections in .json files.

### Database
- Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### Project documentation

Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a **maximum length of 5 minutes**. The content of the video should include:

- A description of the project.
- A demo of the project features.
- You should also highlight some interesting piece of code or something you built that you are very proud of.
- You all should appear in the video and talk about the project.

For recording the video you can use the Zoom recording features while in a call with your peers.

### CODE REVIEW

File: schema.sql [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    CONSTRAINT FK_AUTHORID FOREIGN KEY(ID) REFERENCES Author(ID)
+
+);
+
+CREATE TABLE Author(
+    ID INT CONSTRAINT ID_Author_PK PRIMARY KEY,
+    FIRST_NAME VARCHAR(50),
+    LAST_NAME VARCHAR(50)
+);
+
```
Comment (zainsadaqat) [inline schema.sql:57]: Great job guys for adding all the required tables and setting up the foreign keys. To improve performance, Please also add the INDEXES to the Foreign Keys columns.

#### For Example:
```sql
-- Create an index on the foreign key column
CREATE INDEX idx_items_label_id ON Items(label_id);
```
Screenshot 2023-08-11 at 12 09 14 AM

**Review by zainsadaqat — CHANGES_REQUESTED** [review]
# Required Changes 🔁
## Hello  @MarcoDiaz2000, @tamoorsaeed22, and @ab-noori  👋 🙂

You're doing a great job on your project. We're proud of you. Good job so far! 

It's OK to have bugs in your code. Please, see this code review as a chance for you to learn practical coding skills and learn new tips that can improve your code quality and help you get a better job. 

There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
  ✅  Correct use of Flow
  ✅  Descriptive Pull Request
  ✅  No linter errors
  ✅  Clean code
  ✅  Followed coding best practices
  ✅  Performed pair programming 



## Required Changes ♻️

- [ ] 



## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers, and Happy coding! 👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: schema.sql (hunk 1 above)
Comment (tamoorsaeed22) [reply schema.sql:60]: @zainsadaqat Thank you for your positive feedback! We're delighted to hear that you appreciate the effort we've put into designing and setting up the necessary tables and foreign keys for the database. Your satisfaction is our priority, and we're committed to ensuring that the database operates smoothly and efficiently. We've implemented the recommended INDEXES on the Foreign Key columns, which will significantly enhance the speed and responsiveness of database queries.🙂

**Review by abdulhamiid — APPROVED** [review]
## Hi Team 😀, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**